bootloader/src/old_machine_common_functions.c

```diff
@@ -0,0 +1,109 @@
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include "old_machine_common_functions.h"
+
+char* getRequestDataJson(struct requestData reqData, char* requestFormat, char* systemStr){
+    char* tunnel;
+    if(reqData.tunnel != NULL){
+		size_t tunnelStringSize = sizeof(reqData.tunnel) + (2 * sizeof("\""));
+		tunnel = (char *)malloc(tunnelStringSize);
+        snprintf(tunnel, tunnelStringSize, "%s%s%s", "\"", reqData.tunnel, "\"");
+    } else {
+		tunnel = (char *)malloc(sizeof("false"));
+        strcpy_s(tunnel, sizeof("false"), "false");
+    }
+
+    size_t responseSize = strlen(requestFormat) + strlen(reqData.osVersion) + strlen(reqData.hostname) +
+					      strlen(tunnel) + strlen(reqData.IPstring) + strlen(systemStr);
+
+	if (reqData.glibcVersion != NULL) {
+		responseSize += strlen(reqData.glibcVersion);
+	}
+
+    // Concatenate into string for post data
+    char* buf = malloc(responseSize);
+	if (reqData.glibcVersion != NULL) {
+		snprintf(buf,
+			responseSize,
+			requestFormat,
+			systemStr,
+			reqData.osVersion,
+			reqData.glibcVersion,
+			reqData.hostname,
+			tunnel,
+			reqData.IPstring);
+	}
+	else {
+		snprintf(buf,
+			responseSize,
+			requestFormat,
+			systemStr,
+			reqData.osVersion,
+			reqData.hostname,
+			tunnel,
+			reqData.IPstring);
+	}
+    return buf;
+}
+
+char* concatenate(int size, char** array, const char* joint){
+    size_t jlen = strlen(joint);
+    size_t* lens = malloc(size * sizeof(size_t));
+	if (lens == NULL) {
+		return '\0';
+	}
+    size_t i;
+    size_t total_size = (size-1) * (jlen) + 1;
+    char *result, *p;
+    for(i=0; i<size; ++i){
+		lens[i] = strlen(array[i]);
+        total_size += lens[i];
+    }
+    p = result = malloc(total_size);
```

Comment:
Why is this two variables?

Reply:
Because p is pointer to the end of string and result is pointer to the start of it.